File: Question_2/matrix_mult.py
import time

def matrix_multiply(A, B):
    size = len(A)
    C = [[0] * size for _ in range(size)]
    for i in range(size):
        for j in range(size):
            for k in range(size):
                C[i][j] += A[i][k] * B[k][j]
    return C

sizes = [64, 128, 256, 512, 1024]

for size in sizes:
    A = [[1] * size for _ in range(size)]
    B = [[2] * size for _ in range(size)]

    # System time
    start = time.time()
    C = matrix_multiply(A, B)
    end = time.time()
    system_time = end - start
    print(f"System time for {size}x{size} integer matrix multiplication: {system_time:.10f} seconds")

    # CPU time
    start = time.process_time()
    C = matrix_multiply(A, B)
    end = time.process_time()
    cpu_time = end - start
    print(f"CPU time for {size}x{size} integer matrix multiplication: {cpu_time:.10f} seconds")

File: Question_2/matrix_mult.cpp
#include <iostream>
#include <vector>
#include <ctime>

using namespace std;

void matrix_multiply(const vector<vector<int>>& A, const vector<vector<int>>& B, vector<vector<int>>& C) {
    int N = A.size();
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            C[i][j] = 0;
            for (int k = 0; k < N; ++k) {
                C[i][j] += A[i][k] * B[k][j];
            }
        }
    }
}



int main() {
    vector<int> sizes = {64, 128, 256, 512, 1024};

    for (int size : sizes) {
        vector<vector<int>> A(size, vector<int>(size, 1));
        vector<vector<int>> B(size, vector<int>(size, 2));
        vector<vector<int>> C(size, vector<int>(size, 0));

        timespec start, end;

        // System time
        clock_gettime(CLOCK_MONOTONIC, &start);
        matrix_multiply(A, B, C);
        clock_gettime(CLOCK_MONOTONIC, &end);
        double system_time = (end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec) / 1e9;
        cout << "System time: " << size << "x" << size << " integer matrix multiplication: " << system_time << " seconds" << endl;

        // CPU time
        clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &start);
        matrix_multiply(A, B, C);
        clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &end);
        double cpu_time = (end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec) / 1e9;
        cout << "CPU time:" << size << "x" << size << " integer matrix multiplication: " << cpu_time << " seconds" << endl;
    }

    return 0;
}